File: 446.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:


        # 方法一：差分 + 哈希表 + 动态规划
        # 时间复杂度：O(n^2)
        # 空间复杂度：O(n^2)
        n = len(nums)
        if n < 3:
            return 0

        hash_table = [defaultdict(list) for _ in range(n)]
        #print('hash_table:', hash_table)
        #print(hash_table[0])
        #print('------')
        count = 0

        for i in range(1, n):
            for j in range(i-1, -1, -1):
                diff = nums[i] - nums[j]
                if hash_table[i][diff]:
                    hash_table[i][diff] += (hash_table[j][diff] or 0) + 1
                else:
                    hash_table[i][diff] = (hash_table[j][diff] or 0) + 1                    
                #print(i, j, diff, hash_table)
                if hash_table[j][diff] :
                    count += hash_table[j][diff] 

        return count

File: 446.java
class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        int ans=0;
        Map<Long,Integer>[] mapArr=new HashMap[nums.length];
        for(int i=0;i<nums.length;i++){
            Map<Long,Integer> map=new HashMap<>();
            for(int j=0;j<i;j++){
                long diff=(long)nums[i]-nums[j];
                map.put(diff,map.getOrDefault(diff,0)+mapArr[j].getOrDefault(diff,0)+1);
                ans+=mapArr[j].getOrDefault(diff,0);
            }
            mapArr[i]=map;
        }
        return ans;
    }
}